File: Python3/201 - 400/332 Reconstruct Itinerary.py
# 60ms, 78.67%; 13.5MB, 5.60%
from collections import defaultdict


class Solution:

    def findItinerary(self, tickets: list) -> list:
        self.map, self.res = defaultdict(list), []
        for t1, t2 in sorted(tickets)[::-1]:
            self.map[t1].append(t2)
        self.__visit('JFK')
        return self.res[::-1]

    def __visit(self, x: str) -> None:
        while self.map[x]:
            self.__visit(self.map[x].pop())
        self.res.append(x)

File: 201 - 400/332 Reconstruct Itinerary.cpp
// 24ms, 99.70%; 16.9MB, 32.38%
static int x = []() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    return 0;
}();

class Solution
{
private:
    const string START = "JFK";
    unordered_map<string, multiset<string>> umap;
    vector<string> res;

public:
    vector<string> findItinerary(const vector<pair<string, string>> &tickets)
    {
        for (const auto &p : tickets) umap[p.first].insert(p.second);
        visit(START);
        return vector<string>(res.rbegin(), res.rend());
    }

private:
    void visit(const string &x)
    {
        while (!umap[x].empty()) {
            string next = *umap[x].begin();
            umap[x].erase(umap[x].begin());
            visit(next);
        }
        res.push_back(x);
    }
};
